pkg/migrations/backfill/batcher.go

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package backfill
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/lib/pq"
+	"github.com/xataio/pgroll/pkg/schema"
+)
+
+type batcher struct {
+	table     *schema.Table
+	pkColumn  *schema.Column
+	lastPK    interface{}
+	batchSize int
+}
+
+// updateBatch updates the next batch of rows in the table.
+func (b *batcher) updateBatch(ctx context.Context, conn *sql.DB) error {
+	// Start the transaction for this batch
+	tx, err := conn.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	// Build the query to update the next batch of rows
+	query := b.buildQuery()
+
+	// Execute the query to update the next batch of rows and update the last PK
+	// value for the next batch
+	err = tx.QueryRowContext(ctx, query).Scan(&b.lastPK)
+	if err != nil {
+		return err
+	}
+
+	// Commit the transaction for this batch
+	return tx.Commit()
+}
+
+// buildQuery builds the query used to update the next batch of rows.
+func (b *batcher) buildQuery() string {
+	whereClause := ""
+	if b.lastPK != nil {
+		whereClause = fmt.Sprintf("WHERE %s > %v", pq.QuoteIdentifier(b.pkColumn.Name), b.lastPK)
+	}
+
+	return fmt.Sprintf(`
+    WITH batch AS (
+      SELECT %[1]s FROM %[2]s %[4]s ORDER BY id LIMIT %[3]d FOR UPDATE
```

Comment:
is `ORDER BY id` correct? Is it guaranteed that an id column can always be used?

Reply:
Thanks; that should have been parameterized. I've rebased to fix it.